src/main/java/net/neoforged/neoforge/fluids/potion/PotionBucketItem.java

```diff
@@ -0,0 +1,109 @@
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.ItemUtils;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.TooltipFlag;
+import net.minecraft.world.item.UseAnim;
+import net.minecraft.world.item.alchemy.Potion;
+import net.minecraft.world.item.alchemy.PotionContents;
+import net.minecraft.world.item.alchemy.Potions;
+import net.minecraft.world.level.ItemLike;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.neoforged.neoforge.capabilities.ICapabilityProvider;
+import net.neoforged.neoforge.fluids.capability.IFluidHandlerItem;
+
+public class PotionBucketItem extends Item {
+    public PotionBucketItem(Properties properties) {
+        super(properties);
+    }
+
+    @Override
+    public String getDescriptionId(ItemStack stack) {
+        var potion = stack.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).potion();
+        return Potion.getName(potion, "%s.effect.".formatted(Items.POTION.getDescriptionId()));
+    }
+
+    @Override
+    public void appendHoverText(ItemStack stack, TooltipContext tooltipContext, List<Component> tooltipLines, TooltipFlag tooltipFlag) {
+        var contents = stack.get(DataComponents.POTION_CONTENTS);
+
+        if (contents != null)
+            contents.addPotionTooltip(tooltipLines::add, 1F, tooltipContext.tickRate());
+    }
+
+    @Override
+    public ItemStack getDefaultInstance() {
+        return PotionContents.createItemStack(this, Potions.WATER);
+    }
+
+    @Override
+    public ItemStack finishUsingItem(ItemStack stack, Level level, LivingEntity entity) {
+        var player = entity instanceof Player plr ? plr : null;
+
+        if (player instanceof ServerPlayer sPlayer)
+            CriteriaTriggers.CONSUME_ITEM.trigger(sPlayer, stack);
+
+        if (!level.isClientSide) {
+            stack.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).forEachEffect(effectInstance -> {
+                var effect = effectInstance.getEffect().value();
+
+                if (effect.isInstantenous())
+                    effect.applyInstantenousEffect(player, player, entity, effectInstance.getAmplifier(), 1D);
+                else
+                    entity.addEffect(effectInstance);
+            });
+        }
+
+        if (player != null) {
+            player.awardStat(Stats.ITEM_USED.get(this));
+            stack.consume(1, player);
+        }
+
+        if (player == null || !player.hasInfiniteMaterials()) {
+            if (!stack.isEmpty())
+                return new ItemStack(Items.BUCKET);
+            if (player != null)
+                player.getInventory().add(new ItemStack(Items.BUCKET));
+        }
+
+        entity.gameEvent(GameEvent.DRINK);
+        return stack;
+    }
+
+    @Override
+    public int getUseDuration(ItemStack stack) {
+        return 32;
```

Comment:
Another detail Neo would force that I changed in my implementation, drinking more potion I made take longer. Modders lose the option to choose here. I don't think making this all an API is the right call.